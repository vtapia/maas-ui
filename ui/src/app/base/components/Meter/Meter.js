import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

const DEFAULT_COLORS = ["#007AA6", "#0E8420", "#C7162B", "#F99B11"];

const colorValidator = (props, propName, componentName) => {
  if (props[propName] && !/^#(?:[0-9a-fA-F]{3}){1,2}$/.test(props[propName])) {
    return new Error(
      `Invalid prop ${propName} supplied to ${componentName}. Must be a valid color code.`
    );
  }
};

const Meter = ({
  className,
  data,
  emptyColor = "#D3E4ED",
  labelsClassName,
  max,
  overColor = "#F99B11",
  small = false,
}) => {
  const hasLabels = data.some((datum) => datum.label);
  const valueSum = data.reduce((sum, datum) => sum + datum.value, 0);
  const maximum = max || valueSum;
  const widths = data.map((datum) => (datum.value / maximum) * 100);

  return (
    <div
      className={classNames(small ? "p-meter--small" : "p-meter", className)}
    >
      <div className="p-meter__bar" style={{ backgroundColor: emptyColor }}>
        {valueSum > maximum ? (
          <div
            className="p-meter__filled"
            style={{
              backgroundColor: overColor,
              width: "100%",
            }}
          ></div>
        ) : (
          data.map((datum, i) => (
            <div
              className="p-meter__filled"
              key={`${datum.key}-bar`}
              style={{
                backgroundColor:
                  datum.color || DEFAULT_COLORS[i % DEFAULT_COLORS.length],
                borderRight: `1px solid ${emptyColor}`,
                left: `${widths.reduce(
                  (leftPos, width, j) => (i > j ? leftPos + width : leftPos),
                  0
                )}%`,
                width: `${widths[i]}%`,
              }}
            ></div>
          ))
        )}
      </div>
      {hasLabels && (
        <div className={classNames("p-meter__labels", labelsClassName)}>
          {data.map((datum) => {
            if (!datum.label) {
              return null;
            }
            return (
              <div key={`${datum.key}-label`}>
                {small ? (
                  <small className="u-text--light">{datum.label}</small>
                ) : (
                  datum.label
                )}
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

Meter.propTypes = {
  className: PropTypes.string,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      color: colorValidator,
      key: PropTypes.string.isRequired,
      label: PropTypes.node,
      value: PropTypes.number,
    })
  ).isRequired,
  emptyColor: colorValidator,
  labelsClassName: PropTypes.string,
  max: PropTypes.number,
  overColor: colorValidator,
  small: PropTypes.bool,
};

export default Meter;
