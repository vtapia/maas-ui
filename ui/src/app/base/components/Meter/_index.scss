@mixin Meter {
  $meter-height: $sp-unit * 2;
  $meter-height--small: $sp-unit * 1.5;

  .p-meter {
    margin-bottom: $sp-unit * 1.5;
    padding-top: $spv-inner--x-small;
  }

  .p-meter__bar {
    border-radius: $meter-height;
    height: $meter-height;
    overflow: hidden;
    position: relative;
  }

  .p-meter__labels {
    display: flex;
    justify-content: space-between;
    margin-bottom: -#{$sp-unit * .25};
    padding-top: $sp-unit * 1.25;
  }

  .p-meter__filled {
    height: 100%;
    position: absolute;
    width: 0%;
  }

  .p-meter--small {
    margin-bottom: $sp-unit * 1.75;;
    padding-top: $sp-unit * .75;

    .p-meter__bar {
      border-radius: $meter-height--small;
      height: $meter-height--small;
      margin-bottom: $sp-unit * .75;
    }

    .p-meter__labels {
      margin-bottom: 0;
      padding-top: $sp-unit * .25;
    }
  }
}
