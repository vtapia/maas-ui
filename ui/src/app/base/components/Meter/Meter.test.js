import { shallow } from "enzyme";
import React from "react";

import Meter from "./Meter";

describe("Meter", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Meter
        data={[
          { key: "datum-1", label: "One", value: 1 },
          { key: "datum-2", label: "Two", value: 3 },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can be made small", () => {
    const wrapper = shallow(<Meter data={[]} small />);
    expect(wrapper.find("div").at(0).props().className).toBe("p-meter--small");
  });

  it("can be given labels", () => {
    const wrapper = shallow(
      <Meter
        data={[
          { key: "datum-1", label: "One", value: 1 },
          { key: "datum-2", label: "Two", value: 3 },
        ]}
      />
    );
    expect(wrapper.find(".p-meter__labels div").at(0).text()).toBe("One");
    expect(wrapper.find(".p-meter__labels div").at(1).text()).toBe("Two");
  });

  it("can be given a custom empty colour", () => {
    const wrapper = shallow(<Meter data={[]} emptyColor="#ABC" />);
    expect(wrapper.find(".p-meter__bar").props().style.backgroundColor).toBe(
      "#ABC"
    );
  });

  it("can be given custom bar colours", () => {
    const wrapper = shallow(
      <Meter
        data={[
          { color: "#AAA", key: "aaa", value: 1 },
          { color: "#BBB", key: "bbb", value: 2 },
          { color: "#CCC", key: "ccc", value: 3 },
        ]}
      />
    );
    expect(
      wrapper.find(".p-meter__filled").at(0).props().style.backgroundColor
    ).toBe("#AAA");
    expect(
      wrapper.find(".p-meter__filled").at(1).props().style.backgroundColor
    ).toBe("#BBB");
    expect(
      wrapper.find(".p-meter__filled").at(2).props().style.backgroundColor
    ).toBe("#CCC");
  });

  it("changes colour if values exceed given maximum value", () => {
    const wrapper = shallow(
      <Meter
        data={[{ color: "#ABC", key: "key", value: 100 }]}
        max={10}
        overColor="#DEF"
      />
    );
    expect(wrapper.find(".p-meter__filled").props().style.backgroundColor).toBe(
      "#DEF"
    );
  });

  it("correctly calculates datum widths", () => {
    const wrapper = shallow(
      <Meter
        data={[
          { key: "ten", value: 10 }, // 10/100 = 10%
          { key: "twenty", value: 20 }, // 20/100 = 20%
          { key: "thirty", value: 30 }, // 30/100 = 30%
          { key: "forty", value: 40 }, // 40/100 = 40%
        ]}
      />
    );
    expect(wrapper.find(".p-meter__filled").at(0).props().style.width).toBe(
      "10%"
    );
    expect(wrapper.find(".p-meter__filled").at(1).props().style.width).toBe(
      "20%"
    );
    expect(wrapper.find(".p-meter__filled").at(2).props().style.width).toBe(
      "30%"
    );
    expect(wrapper.find(".p-meter__filled").at(3).props().style.width).toBe(
      "40%"
    );
  });

  it("correctly calculates datum positions", () => {
    const wrapper = shallow(
      <Meter
        data={[
          { key: "ten", value: 10 }, // 1st = 0%
          { key: "twenty", value: 20 }, // 2nd = 1st width = 10%
          { key: "thirty", value: 30 }, // 3rd = 1st + 2nd width = 30%
          { key: "forty", value: 40 }, // 4th = 1st + 2nd + 3rd width = 60%
        ]}
      />
    );
    expect(wrapper.find(".p-meter__filled").at(0).props().style.left).toBe(
      "0%"
    );
    expect(wrapper.find(".p-meter__filled").at(1).props().style.left).toBe(
      "10%"
    );
    expect(wrapper.find(".p-meter__filled").at(2).props().style.left).toBe(
      "30%"
    );
    expect(wrapper.find(".p-meter__filled").at(3).props().style.left).toBe(
      "60%"
    );
  });
});
