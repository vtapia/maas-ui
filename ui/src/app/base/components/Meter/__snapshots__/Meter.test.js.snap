// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meter renders 1`] = `
<div
  className="p-meter"
>
  <div
    className="p-meter__bar"
    style={
      Object {
        "backgroundColor": "#D3E4ED",
      }
    }
  >
    <div
      className="p-meter__filled"
      key="datum-1-bar"
      style={
        Object {
          "backgroundColor": "#007AA6",
          "borderRight": "1px solid #D3E4ED",
          "left": "0%",
          "width": "25%",
        }
      }
    />
    <div
      className="p-meter__filled"
      key="datum-2-bar"
      style={
        Object {
          "backgroundColor": "#0E8420",
          "borderRight": "1px solid #D3E4ED",
          "left": "25%",
          "width": "75%",
        }
      }
    />
  </div>
  <div
    className="p-meter__labels"
  >
    <div
      key="datum-1-label"
    >
      One
    </div>
    <div
      key="datum-2-label"
    >
      Two
    </div>
  </div>
</div>
`;
